Add simulated base IMU with noise terms

nrotella opened the pull request:
## Description

This small PR adds a simulated base IMU which provides 3d gyroscope and 3d accelerometer measurements including noise terms. The gyroscope data is computed as the base angular velocity in base frame plus noise, while the accelerometer data as the base acceleration (differentiated numerically from base linear velocity) with negative gravity added, rotated into base frame plus noise.

The noise terms include those most commonly observed in MEMS sensors: 3d gyroscope bias, 3d gyroscope thermal noise, 3d accelerometer bias, and 3d accelerometer noise. The bias terms are modeled as Brownian motion (random walk), computed by integrating Gaussian white noise having known standard deviation (discretized using the simulation timestep). The thermal noise is modeled as additive Gaussian white noise having known standard deviation (also discretized as above). The continuous noise term standard deviations are set to nominal values based on an older IMU (Microstrain 3DM-GX3-25) using in Rotella 2014 as an example.

I had intended to write a more general simulated IMU class which could be associated with any link and with a relative pose offset to that link (as we discussed) but wanted to get something implemented for now, so I settled on a simple base IMU assumed to be pose-aligned with the base link.

## How I Tested

The COM control task in blmc_controllers was modified to perform a fast sinusoidal squatting motion. The simulated IMU getter functions added in this PR were used to obtain sensor data, which was recorded to a pkl file using the DataCollector class. Data was analyzed in RAI using the built-in signal differentation functionality to confirm that the numerical acceleration matches differentiated velocity as expected, as well as to verify that the noise terms are functioning properly. 

Note that the effect of the biases cannot be seen clearly unless the continuous noise values are increased drastically; IMU bias terms grow very slowly over long periods of operation.

nrotella committed:
Add base acceleration computation and getter

Added a generic function to compute numerical quantities from
simulation results using the bullet env timestep, namely the base
linear and angular acceleration. These were verified in the solo12
simulation COM wrench control task for a vertical squatting motion.

nrotella committed:
Add simulated base IMU with noise

Added a simulated base IMU with noise terms (bias and thermal noise)
using the base velocity and numerically-computed acceleration to
provide gyroscope and accelerometer measurements, respectively.

MaximilienNaveau reviewed (approved):
This seems fine, I guess we are sensing at the center of the base.
Maybe we could later on introduce the fact that the imu could be placed at a different location?

nrotella commented:
Yeah we can add this, I can push it to this PR if you want before it gets
merged. I can add it this evening I think.

Assuming the IMU stays on the base (for now I don't see why it would need
to be on any other link) then adding a pose offset modifies the
acceleration according to

$a_{IMU} = a_{base} + \alpha \cross r_{IMU, base} + \omega \cross \omega
\cross r_{IMU, base}$

where $r_{IMU, base}$ is the pose translational part, and then both the
angular velocity and modified linear acceleration are rotated by the pose
rotational part.

Sorry if my latex is wrong here, somehow the plugin to render in GitHub no
longer works for me.

On Thu, Jan 21, 2021, 7:58 AM Naveau <notifications@github.com> wrote:

> *@MaximilienNaveau* approved this pull request.
>
> This seems fine, I guess we are sensing at the center of the base.
> Maybe we could later on introduce the fact that the imu could be placed at
> a different location?
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/machines-in-motion/bullet_utils/pull/2#pullrequestreview-573352820>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACURXNJ5U46HRUKUI5V6AZLS3AXJNANCNFSM4WMAFN7Q>
> .
>

jviereck commented on src/bullet_utils/wrapper.py:
```diff
@@ -33,6 +34,22 @@ def __init__(
+        self.base_imu_accel_bias = np.zeros((3, 1))
+        self.base_imu_gyro_bias = np.zeros((3, 1))
+        self.base_imu_accel_thermal = np.zeros((3, 1))
+        self.base_imu_gyro_thermal = np.zeros((3, 1))
```
Is there a reason you are allocating arrays of size `(3, 1)`? Usually we try to go with one-dimensional arrays, so in this case only of size `np.zeros(3)`.

nrotella replied:
I was having some numpy issues with matrix-vector products in another line, so I explicitly used column vectors everywhere to try to fix it. To be honest, though, I'm not great with numpy, so maybe this was unnecessary. The idea of using one-dimensional arrays instead of explicit column (3,1) or row (1,3) vectors is weird to me coming from Eigen. I also don't know what best practice is for numpy, or if this is more a preference thing?

I can definitely try to rewrite these commits using 1d arrays.

jviereck commented:
Hi Nick, 

thanks for your PR. This looks good overall besides the comment about using arrays of shape (3, 1) over just (3) (see the review comment).

Having the ability to offset the IMU from the base would be good. On solo12, the IMU is not at the base but offset by roughly 10 cm towards the front.

nrotella commented:
> Hi Nick,
> 
> thanks for your PR. This looks good overall besides the comment about using arrays of shape (3, 1) over just (3) (see the review comment).
> 
> Having the ability to offset the IMU from the base would be good. On solo12, the IMU is not at the base but offset by roughly 10 cm towards the front.

Thanks for taking a look at these changes. Is the IMU on solo12 mounted on the base but offset some amount? Is there a rotation offset as well? A diagram would help, but for now, I can add a generic 6 dof pose offset (assuming the IMU will always be on the base rigid body and not on another link).

jviereck commented:
> Is the IMU on solo12 mounted on the base but offset some amount? Is there a rotation offset as well?

For solo12, the IMU is mounted on the base but with some fixed offset. For solo12 there is no rotation offset but for solo8 I think there is.

Enclosed you find a picture of the robot. There where I am pointing the IMU is mounted from below.

![PXL_20210121_153939913](https://user-images.githubusercontent.com/191719/105374421-b2bfad80-5bd5-11eb-82d0-7fc194d5f16a.jpg)

jviereck commented:
Asking Felix about the IMU position, a rough estimate is for solo12

> Its about 116mm forward from the robot origin and 17mm upwards.

nrotella commented:
> Asking Felix about the IMU position, a rough estimate is for solo12
> 
> > Its about 116mm forward from the robot origin and 17mm upwards.

Great thanks for the specs. Can you also remind me what the coordinate frame for the base frame is? Not sure if there's an easy way to draw joint or link frames for visualization purposes in bullet like we had in SL?

jviereck commented:
We have a drawing with the frames for solo12 here:

https://github.com/open-dynamic-robot-initiative/open_robot_actuator_hardware/blob/master/mechanics/quadruped_robot_12dof_v1/images/solo12_coordinate_systems.png

Hope that helps.

I haven't seen the ability to visualize frames in bullet like it was in SL yet.

ahmadgazar commented:
> Sorry if my latex is wrong here, somehow the plugin to render in GitHub no longer works for me.

It just might be a bit different from latex sometimes with mathjax. For example, in your comment change /cross with /times to render.

nrotella commented:
Ah, my mistake, it should be \times for mathjax as well. I really need to
get my plugin working again, it was super useful. Thanks!

On Thu, Jan 21, 2021, 10:36 AM Ahmad Gazar <notifications@github.com> wrote:

> Sorry if my latex is wrong here, somehow the plugin to render in GitHub no
> longer works for me.
>
> It just might be a bit different from latex sometimes with mathjax. For
> example, in your comment change /cross with /times to render.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/machines-in-motion/bullet_utils/pull/2#issuecomment-764775724>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACURXNKXCQKWTGF6GT44RQDS3BJZLANCNFSM4WMAFN7Q>
> .
>

nrotella commented:
Thanks! This functionality is very helpful to have IMO. Not sure what it
would take to add it to our bullet wrapper, but I guess it's more of a core
bullet feature at a lower level.

On Thu, Jan 21, 2021, 10:36 AM Julian Viereck <notifications@github.com>
wrote:

> We have a drawing with the frames for solo12 here:
>
>
> https://github.com/open-dynamic-robot-initiative/open_robot_actuator_hardware/blob/master/mechanics/quadruped_robot_12dof_v1/images/solo12_coordinate_systems.png
>
> Hope that helps.
>
> I haven't seen the ability to visualize frames in bullet like it was in SL
> yet.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/machines-in-motion/bullet_utils/pull/2#issuecomment-764775584>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACURXNPC3LSUCBBOOJX5TPDS3BJYZANCNFSM4WMAFN7Q>
> .
>

nrotella committed:
Fix numpy array sizes to be 1d

nrotella committed:
Add IMU pose offset relative to base

Added a simple pose offset for the IMU relative to the based, used to
compute the IMU angular velocity and linear acceleration. The offset
orientation is the identity, and the offset translation is set based on
specified IMU location (0.116m forward, 0.017m upward). Tested only
to verify that the identity rotation and specified translation appear
to work correctly in transforming IMU data.

nrotella commented:
I pushed a simple modification to offset the IMU relative to the base as [described in an earlier comment](https://github.com/machines-in-motion/bullet_utils/pull/2#issuecomment-764668512) I used the relative position measurements @jviereck provided, but assumed the relative orientation is the identity. The current IMU pose offset thus only affects the IMU measured acceleration.

I have no way of testing that the offset IMU data is correct, but I checked that the data changes how I would expect given a large offset. If you find an issue with this in testing, please let me know, but it looks ok to me.

MaximilienNaveau commented:
I am merging if we find a problem we will create another issue/PR
